```{r}
# Load the packages
library(arrow)
library(tidyr)
library(cowplot)
library(ggplot2)
library(dplyr)
library(stringr)

# File path the feather file containing structural data editted in python script.
file_path <- "C:\\Users\\Adam Martin\\OneDrive - University of Dundee\\Honours Project\\231006a-AM-honours-project-aps\\varalign-data-PF00026\\feather\\PF00026.22_swissprot.sto_prointvar_structure_table.p.gz.feather"

# Read the Feather file
structure <- arrow::read_feather(file_path)
```


```{r}
# Works out the number of different proteins in structure file
n_strucs <- length(unique(structure$PDB_dbAccessionId_A))
n_strucs
# Works out the number of different proteins in structure file
n_prots <- length(unique(structure$UniProt_dbAccessionId_A))
n_prots
count_prot <- table(structure$SOURCE_ID_A)


# Cuts down the names of the proteins
cleaned_proteins <- sub("/.*", "", structure$SOURCE_ID_A)
human_proteins <- cleaned_proteins[grep("HUMAN", cleaned_proteins)]
human_proteins <- human_proteins[grep("HUMAN", human_proteins)]

# Human only proteins
human_proteins_for_writing <- sub("_.*", "", human_proteins)

```


```{r}

# Input the csv file that contains all the general information of Aspartyl proteases from VarAlign
aspartyl_proteases <- read.csv("C:/Users/Adam Martin/OneDrive - University of Dundee/Honours Project/231006a-AM-honours-project-aps/varalign-data-PF00026/csv/aspartyl_proteases.csv")


#Filters the table for columns with over 5 occupancy. This is the same as > 50% occupancy as there are 10 human sequences 
aspartyl_proteases <- aspartyl_proteases %>%
  filter(occupancy > 5)


# Adds a new index to the table of with the columns with > 50% occupancy
aspartyl_proteases <- aspartyl_proteases %>% mutate(dex = row_number())
```






```{r}
# Assignment of the of CMDs and UMDs according to MacGowan et al
aspartyl_proteases <- aspartyl_proteases %>%
  mutate(key.y = case_when(
    mes_p > 0.5  ~ "No significance",
    shenkin_nrank < 0.25 & mes_or < 1 ~ "CMD",
    shenkin_nrank > 0.75 & mes_or < 1 ~ "UMD",
    shenkin_nrank < 0.25 & mes_or > 1 ~ "No significance",
    shenkin_nrank > 0.75 & mes_or > 1 ~"No significance",
    TRUE ~ "No significance"  # default color for other cases
))

# Labeled the active site residues, if they were classified as Not significant
active_site_column <- c(40,41,42, 475,476,477)
for (column_number in active_site_column) {
aspartyl_proteases <- aspartyl_proteases %>%
  mutate(key.y = case_when(
    column == column_number & key.y == "No significance" ~ "Active site",
    TRUE ~ key.y
  )) }

# Reorders the assignments to change it in the graphs
legend_order <- c( "CMD", "UMD", "No significance", "Active site")
aspartyl_proteases$key <- factor(aspartyl_proteases$key, levels = legend_order)

# Stores the active site information within a dataframe
active_site <- which(aspartyl_proteases$key == "Active site") 
active_site


# Dataframe of active site, CMDs and UMDs 
CMD_UMD_Active <- aspartyl_proteases %>%
  filter(key %in% c("UMD", "CMD", "Active site"))%>%
  select(column, key, dex)
CMD_UMD_Active
```

###### MES graph, active site visuable ######
```{r}
# Function used to plot the Missense enrichment graph.
# Plots each column, with the ranked normalised shenkin against the odds ratio.

number_rows <- nrow(aspartyl_proteases)

your_plot <-ggplot(aspartyl_proteases, aes(x = shenkin_nrank, y = mes_or, color = key, alpha = key)) +
  geom_point() +
  #geom_text(data = aspartyl_proteases, aes(label = column), hjust = -0.2, vjust = 0.5, size = 4) +
  labs(x = "Normalised conservation score", y = "Missense score (human variation)") +
  geom_hline(yintercept = 1, linetype = "dashed", color = "black") +
  geom_vline(xintercept = 0.25, linetype = "dashed", color = "black") +
  geom_vline(xintercept = 0.75, linetype = "dashed", color = "black") + 
  #scale_color_manual(values = c("grey" = "grey", "blue" = "blue", "red" = "red", "purple" = "purple", "black"="black" )) + 
  scale_color_manual(values = c("Active site" = "purple", "No significance" = "grey", "black" = "black", "UMD" = "red", "CMD" = "blue", "UME" = "yellow" )) +
  scale_alpha_manual(values = c("Active site" = 0.8, "No significance" = 0.4, "CMD" = 0.8, "UMD" = 0.8)) +
  scale_y_log10()

your_plot


ggsave("C:/Users/Adam Martin/OneDrive - University of Dundee/Documents/Honour Project Dump/MES_shenkin_GnomadV3.png", plot = your_plot, width = 6, height = 4, units = "in")


```







  
        
        


######Protein, ligand interactions, Active site visuable ######




```{r}

# Fills in the gaps within the protein and ligand interaction columns
aspartyl_proteases$PDB_protein_ligand_interactions <- ifelse(is.na(aspartyl_proteases$PDB_protein_ligand_interactions), 0, aspartyl_proteases$PDB_protein_ligand_interactions)
aspartyl_proteases$PDB_protein_protein_interactions <- ifelse(is.na(aspartyl_proteases$PDB_protein_protein_interactions), 0, aspartyl_proteases$PDB_protein_protein_interactions)


number_rows <- nrow(aspartyl_proteases)

# Function used to plot the protein and ligand graphs, colouring the active site, CMDs and UMDs
ligand_or_protein_graph <- function(data, y_column, key, yaxis, save_path) {
  your_plot <- ggplot(data, aes(x = dex, y = {{ y_column }}, fill = {{ key }})) +
    geom_bar(stat = "identity") +
    coord_cartesian(ylim = c(0, 16), xlim = c(0, number_rows)) +
    scale_y_continuous(breaks = seq(0, 16, by = 2)) +
    scale_x_continuous(breaks = seq(0, number_rows, by = 25)) +
    labs(x = ifelse(key == "Protein", "Alignment Column", ""), y = yaxis) +
    scale_fill_manual(values = c("UMD" = "red", "CMD" = "blue", "No significance" = "grey", "Active site" = "purple")) +
    theme(legend.position = ifelse( yaxis== "Protein", "bottom", "none"), legend.justification = "center",
          legend.title = element_blank(), plot.margin = margin(0, 0, 0, 0))
  print(your_plot)
  
}

# Plotting the ligand interaction graph using function above
plot1 <- ligand_or_protein_graph(aspartyl_proteases, aspartyl_proteases$PDB_protein_ligand_interactions,aspartyl_proteases$key, "Ligand",  "C:/Users/Adam Martin/OneDrive - University of Dundee/Documents/Honour Project Dump/ligand_interaction.png")

# Plotting the protein interaction graph using function above
plot2 <- ligand_or_protein_graph(aspartyl_proteases, aspartyl_proteases$PDB_protein_protein_interactions,aspartyl_proteases$key, "Protein",  "C:/Users/Adam Martin/OneDrive - University of Dundee/Documents/Honour Project Dump/protein_interaction.png")




# Used to combine the 2 plots above, protein and ligand interactions to make 1 figure used for the thesis write up
plot_list <- list(plot1, plot2)

combined_plot <- plot_grid(
  plotlist = plot_list,
  ncol = 1,
  widths = unit(30, "inches"),
  rel_heights = c(2, 1.4)
)


save_path <- "C:/Users/Adam Martin/OneDrive - University of Dundee/Documents/Honour Project Dump/ligand_protein_interaction.png"
print(combined_plot)
ggsave(save_path, plot = combined_plot, width = 30, height = 15, units = "in")
```










```{r}
number_rows <- nrow(aspartyl_proteases)

# Function used to plot each secondary structure and amino acid position. Plotted for each of the CMD, UMD and active site highlighted. 
create_custom_plot <- function(selected_columns, index_of_UMD_CMD_active_site, title, key, i) {
  selected_columns$index <- seq_len(nrow(selected_columns))
  df_long <- gather(selected_columns, key = "Variable", value = "Value", -index)
  df_combined <- cbind(df_long, keys = key)
  
  p <- ggplot(df_combined, aes(x = index, y = Value, fill = Variable, alpha = as.factor(index %in% index_of_UMD_CMD_active_site))) +
    geom_bar(stat = "identity") +
    #theme_minimal() +
    labs(x = ifelse(i == "Active site", "Alignment Column", ""), y = i, fill = "") +
    coord_cartesian(ylim = c(0, 16), xlim = c(0, number_rows)) +
    scale_y_continuous(breaks = seq(0, 16, by = 4)) +
    scale_x_continuous(breaks = seq(0, number_rows, by = 25)) +
    scale_alpha_manual(values = c("TRUE" = 5, "FALSE" = 0.2), guide = "none") +
    theme(legend.position = ifelse(i == "Active site", "bottom", "none"), legend.justification ="center", plot.margin = margin(0, 0, 0, 0),      plot.title = element_text(hjust = 0.5))  # Center the title

  if (i == "CMD") {
    p <- p + ggtitle(title)  # Add your desired title
  }
  print(p)
  return(p)
}
```


```{r}
# Function used to read in a feather file from Jupyter notebook
read_feather_file <- function(file_path) {
  data <- arrow::read_feather(file_path)
  return(data)
}

# Function uses the plot function above 
# The graphs are then combined to be used in the thesis write up
create_plot <- function(selected_columns, CMD_UMD_Active, title, key) {
  plot_list <- list()

  for (i in c("CMD", "UMD", "Active site")) {
    CMD_UMD_Active_filter <- CMD_UMD_Active %>% filter(key == i)
    plot <- create_custom_plot(selected_columns, CMD_UMD_Active_filter$dex,  title, key, i)
    plot_list[[i]] <- plot
  }

  combined_plot <- plot_grid(plotlist = plot_list, ncol = 1,  widths = unit(8.27, "inches"), rel_heights = c(1.6, 1.3, 1))
  print(combined_plot)
  return(combined_plot)
}
```


```{r}
# Active site, CMD and UMD position dataframe 
CMD_UMD_Active <- aspartyl_proteases %>%
  filter(key %in% c("UMD", "CMD", "Active site"))%>%
  select(column, key, dex)
CMD_UMD_Active


# File path
file_path <- "C:/Users/Adam Martin/OneDrive - University of Dundee/Honours Project/231006a-AM-honours-project-aps/varalign-data-PF00026/feather/"


# The RSA feather file used to generate the graphs for thesis, using the 2 functions above
rsa_class <- read_feather_file(paste0(file_path, "rsa_class_df.feather"))
rsa_class <- rsa_class %>% rename(column = "__index_level_0__")
rsa_class <- left_join(aspartyl_proteases, rsa_class, by = "column")
selected_columns <- rsa_class[c("core", "part", "surf")]
plot<-create_plot(selected_columns, CMD_UMD_Active, "Amino acid position", "RSA")
save_path <-"C:/Users/Adam Martin/OneDrive - University of Dundee/Documents/Honour Project Dump/RSA_GnomadV3.png"
ggsave(save_path, plot = plot, width = 6, height = 4, units = "in")

# The Secondary structure feather file used to generate the graphs for thesis, using the 2 functions above
ss_class <- read_feather_file(paste0(file_path,"ss_class_df.feather"))
ss_class <- ss_class %>% rename(column = "__index_level_0__")
ss_class <- left_join(aspartyl_proteases, ss_class, by = "column")
selected_columns <- ss_class[c("helix", "coil", "strand")]
plot<-create_plot(selected_columns, CMD_UMD_Active, "Secondary structure", "SS")
save_path <-"C:/Users/Adam Martin/OneDrive - University of Dundee/Documents/Honour Project Dump/SS_GnomadV3.png"
ggsave(save_path, plot = plot, width = 6, height = 4, units = "in")
```



#create_custom_plot <- function(selected_columns, index_of_UMD_CMD_active_site, title, key, i) {
#  selected_columns$index <- seq_len(nrow(selected_columns))
#  df_long <- gather(selected_columns, key = "Variable", value = "Value", -index)
#  df_combined <- cbind(df_long, keys = key)
  
#  p <- ggplot(df_combined, aes(x = index, y = Value, fill = Variable, alpha = as.factor(index %in% index_of_UMD_CMD_active_site))) +#
#    geom_bar(stat = "identity") +
#    labs(x = ifelse(i == "Active site", "Alignment Column", ""), y = i, fill = "") +
#    coord_cartesian(ylim = c(0, 16), xlim = c(0, number_rows)) +
#    scale_y_continuous(breaks = seq(0, 16, by = 4)) +
#    scale_x_continuous(breaks = seq(0, number_rows, by = 25)) +
#    scale_alpha_manual(values = c("TRUE" = 2, "FALSE" = 0.25), guide = "none") +
#    theme(legend.position = ifelse(i == "Active site", "bottom", "none"),
#          legend.justification ="center",
#          plot.margin = margin(0, 0, 0, 0),
#          axis.text.x = ifelse(i == "Active site", element_text(), element_blank())
#    )
#  print(p)
#  return(p)
#}






##### Gnomad v4 testing ######
```{r}
# Gets a list of the human proteins used within the analysis and output of VarAlign
unique_protein_names <-  unique(structure$SOURCE_ID_A)
human_proteins_accessions <- unique_protein_names[grep("HUMAN", unique_protein_names)]
human_proteins <-  sub("_HUMAN", "", human_proteins_accessions)
human_proteins <-  sub("/.*$", "", human_proteins)
human_proteins



# Gets the gnomAD accession numbers attached with the different human proteins
protein_to_gnomad_accession <- structure %>%
  filter(!duplicated(Ensembl_dbAccessionId_A))
protein_to_gnomad_accession <- protein_to_gnomad_accession %>%
  filter(SOURCE_ID_A %in% human_proteins_accessions)

# Uses the dataframe. The accession numbers where used to search gnomAD v4 and download the files
protein_to_gnomad_accession[,c("SOURCE_ID_A", "Ensembl_dbAccessionId_A")]
```




```{r}
# THis is used to generate a table that merges the structure table and aspartyl protease table that were outputs from VarAlign 
colnames(structure)[colnames(structure) == "Alignment_column_A"] <- "column"
mini_aspartyl_proteases <- aspartyl_proteases  %>%
  #filter(key %in% c("CMD", "UMD", "Active site"))%>%
  select(key, column, dex)

merged_structure <- merge(structure, mini_aspartyl_proteases, by = "column")

merged_structure_dictionary_human <- merged_structure %>%
  filter( SOURCE_ID_A %in% human_proteins_accessions)  %>%
  select(SOURCE_ID_A, key, column, UniProt_dbResNum_A, dex)%>%
  distinct()
```


```{r}
# File path to the feather file generated using Jupyter notebook of the varaints table from VarAlign output
file_path <- "C:/Users/Adam Martin/OneDrive - University of Dundee/Honours Project/231006a-AM-honours-project-aps/varalign-data-PF00026/feather/PF00026.22_swissprot.sto_variants.p.gz.feather"
# Read the Feather file
variants <- arrow::read_feather(file_path)
colnames(variants)[colnames(variants) == "alignment_column"] <- "column"
```



```{r}
# Generates a dataframe filtered for only the CMD, UMD and active site positions from VarAlign output, will be used for the analysis later on 
original_alignment_important_human_variants <- aspartyl_proteases %>%
  filter(key %in% c("CMD", "UMD", "Active site"))
original_alignment_important_human_variants

original_alignment_important_human_variants$missense_variant

```


```{r}
# Function used to input a csv files of the gnomAD v4 files, it also break the HGVS.consequence column into what the amino acid was, the positions within the protein and the resulting amino acid of the missense mutation
read_csv_file <- function(file_path) {
  data <- read.csv(file_path)
  data <- data %>% filter(VEP.Annotation == "missense_variant") #### Added in VEP.Annotation selecting for only missense varaints
  data$HGVS.Consequence <- sub("^p\\.", "", data$HGVS.Consequence)
  data$Amino_Acid_original <- str_extract(data$HGVS.Consequence, "[A-Za-z]+")
  data$Position <- as.numeric(str_extract(data$HGVS.Consequence, "\\d+"))
  data$Amino_Acid_missense_mutation <- str_extract(data$HGVS.Consequence, "[A-Za-z]{3}$")
  
  
  return(data)
}

file_path <- "C:/Users/Adam Martin/OneDrive - University of Dundee/Honours Project/231006a-AM-honours-project-aps/varalign-data-PF00026/csv/gnomadv4/"


protein_to_gnomad_accession[,c("SOURCE_ID_A", "Ensembl_dbAccessionId_A")]
# CATE was found on GnomAD v4 may not have been present in gnomAD v2

# Uses the function above for each of the gnomAD v4 files 
RENI_HUMAN <- read_csv_file(paste0(file_path,"gnomAD_v4.0.0_ENSG00000143839_2023_12_29_21_14_25.csv"))
CATD_HUMAN <- read_csv_file(paste0(file_path,"gnomAD_v4.0.0_ENSG00000117984_2023_12_29_21_58_37.csv"))  
PEPA4_HUMAN <- read_csv_file(paste0(file_path,"gnomAD_v4.0.0_ENSG00000229183_2024_01_14_22_36_30.csv"))
CATE_HUMAN <- read_csv_file(paste0(file_path,"gnomAD_v4.0.0_ENSG00000196188_2024_01_14_22_23_16.csv")) 
PEPC_HUMAN <- read_csv_file(paste0(file_path,"gnomAD_v4.0.0_ENSG00000096088_2023_12_29_21_59_00.csv")) 
BACE1_HUMAN <- read_csv_file(paste0(file_path,"gnomAD_v4.0.0_ENSG00000186318_2023_12_29_21_59_20.csv"))
BACE2_HUMAN <- read_csv_file(paste0(file_path,"gnomAD_v4.0.0_ENSG00000182240_2023_12_29_21_59_42.csv"))
```






```{r}
# Function that used for the dataframe generated in cell above for each of the gnomAD v4 files. It takes in the dataframe and counts the number of each different missense variants at each alignemnt column.
protein_gnomadv4_matching <- function(merged_structure_dictionary_human, protein, protein_df) {
  alignment <- merged_structure_dictionary_human %>% filter( SOURCE_ID_A == protein)
  df <- data.frame(ID = numeric(0), Name = character(0))
  
 for (i in alignment$UniProt_dbResNum_A) {
    matching_rows <- filter(protein_df, Position == i)
    
    if (nrow(matching_rows) > 0) {
      matching_rows$protien_position <- i
      alignment_column_key <- alignment  %>%
      filter(UniProt_dbResNum_A == i)    %>%
      select(column , key)
      combined_df <- cbind(matching_rows, alignment_column_key)
      df <- rbind(df, combined_df)
    }
  }
  df <- df %>%
  mutate(Gene = protein) %>%
  select(Gene, everything())
  df <- df %>%
    filter("VEP.Annotation"== "missense_variant")
  
return(df)#[, c( "Gene", "HGVS.Consequence", "Allele.Count", "protien_position", "column", "key")])
  
}

# uses function above, counting the number of variants within each column
RENI_HUMAN_sites <- protein_gnomadv4_matching(merged_structure_dictionary_human, "RENI_HUMAN/85-405", RENI_HUMAN)
CATD_HUMAN_sites <- protein_gnomadv4_matching( merged_structure_dictionary_human, "CATD_HUMAN/78-409", CATD_HUMAN)
PEPA4_HUMAN_sites <- protein_gnomadv4_matching( merged_structure_dictionary_human, "PEPA4_HUMAN/75-387", PEPA4_HUMAN)
CATE_HUMAN_sites <- protein_gnomadv4_matching( merged_structure_dictionary_human, "CATE_HUMAN/77-399", CATE_HUMAN)
PEPC_HUMAN_sites <- protein_gnomadv4_matching( merged_structure_dictionary_human, "PEPC_HUMAN/72-387", PEPC_HUMAN)
BACE1_HUMAN_sites <- protein_gnomadv4_matching( merged_structure_dictionary_human, "BACE1_HUMAN/74-418", BACE1_HUMAN)
BACE2_HUMAN_sites <- protein_gnomadv4_matching( merged_structure_dictionary_human, "BACE2_HUMAN/91-431", BACE2_HUMAN)



merged_structure_dictionary_human

# Combines each of the dataframes above. Its then used in the same manner as the original VarAlign output
All_HUMAN_sites <- rbind(RENI_HUMAN_sites, CATD_HUMAN_sites, PEPA4_HUMAN_sites, CATE_HUMAN_sites, PEPC_HUMAN_sites, BACE1_HUMAN_sites, BACE2_HUMAN_sites)
All_HUMAN_sites <- All_HUMAN_sites%>%
  filter(Filters...exomes == "PASS")#, Filters...genomes == "PASS")  #Only if it passes the exome filter
```



```{r}
# This cell is used for calculating the Missense enrichment score. This cell is based of the Ankyrin paper code found here "https://github.com/bartongroup/JSU_ANK_analysis"
aspartyl_proteases$occupancy


# Create a 2x2 contingency table
# Calculates the total human occupancy within the alignment and total number of missense variants
total_occupancy <- sum(aspartyl_proteases$occupancy)
total_missense <- sum(All_HUMAN_sites$Allele.Count)


# Opens multiple dataframes used for later analysis
df_fisher_MES <- data.frame(ID = numeric(0), Name = character(0))
df <- data.frame(ID = numeric(0), Name = character(0))
my_list <- list()
missense_missense <- data.frame(column = numeric(),
                    gnomADv3 = numeric(),
                    gnomADv4 = numeric())


# Used to calculate the missense enrichment score, using odds ratio. A 2x2 matrix comparing the occupancy of the column to the remainging occupancy with the number of missense variants within a column to the remaining
# missense variants.
# Saved into a table used for later analysis
for (i in unique(merged_structure_dictionary_human$column)) {
  missense_at_pos <- subset(All_HUMAN_sites, column == i)
  #sum_missense_at_position <- sum(subset(All_HUMAN_sites, column == i)$Allele.Count)    #Change sum to average see the difference 
  sum_missense_at_position <- nrow(missense_at_pos)
  total_missense <- nrow(All_HUMAN_sites)
  occupancy <- aspartyl_proteases%>%
    filter(column == i) %>%
    select(occupancy)
  occupancy <- occupancy$occupancy
  
  data <- matrix(c(sum_missense_at_position,(total_missense-sum_missense_at_position),occupancy,(total_occupancy-occupancy) ), nrow = 2)
  fisher_result <- fisher.test(data)
  combined_df <- cbind(i, fisher_result$p.value, fisher_result$conf.int[1], fisher_result$conf.int[2], fisher_result$estimate)
  df_fisher_MES <- rbind(df_fisher_MES, combined_df)
  
  
 
   new_row <- data.frame(column = i,
                    gnomADv3 = aspartyl_proteases %>% filter(column == i) %>% select(X_missense_all) ,
                    gnomADv4 = sum_missense_at_position)
    missense_missense <- rbind(missense_missense, new_row)


}

# Renaming some of the columns
merged_missense_missense_aspartyl <- merge(missense_missense, aspartyl_proteases, by = "column")

colnames(df_fisher_MES)[colnames(df_fisher_MES) == "i"] <- "column"
colnames(df_fisher_MES)[colnames(df_fisher_MES) == "V2"] <- "mes_p"
colnames(df_fisher_MES)[colnames(df_fisher_MES) == "V3"] <- "conf.int_lower"
colnames(df_fisher_MES)[colnames(df_fisher_MES) == "V4"] <- "conf.int_higher"
colnames(df_fisher_MES)[colnames(df_fisher_MES) == "V5"] <- "mes_or"
df_fisher_MES <- distinct(df_fisher_MES)
print(df_fisher_MES)
```


```{r}
# Output the gnomAD v4 csv
file_path <- "C:/Users/Adam Martin/OneDrive - University of Dundee/Documents/Honour Project Dump/missense_comparison_4.csv"

# Export the data frame to CSV
write.csv(missense_missense, file = file_path, row.names = FALSE)

# Print a message indicating success
cat("Data frame exported to CSV:", file_path, "\n")
```

```{r}
# Adding the ranked normalised shenkin from the original VarAlign output to the gnomAD v4 dataframe 
n_rank <- aspartyl_proteases[, c("column", "shenkin_nrank")]
n_rank


merged_df <- merge(df_fisher_MES, n_rank, by = "column")
merged_df
```

```{r}
# Assigning CMD and UMD as before to locate CMDs and UMDs 
merged_df <- merged_df %>%
  mutate(key_GnomadV4 = case_when(
    
    mes_p > 0.5  ~ "No significance",
    shenkin_nrank < 0.25 & mes_or < 1 ~ "CMD",
    shenkin_nrank > 0.75 & mes_or < 1 ~ "UMD",
    shenkin_nrank < 0.25 & mes_or > 1 ~ "No significance",
    shenkin_nrank > 0.75 & mes_or > 1 ~"No significance",
    TRUE ~ "No significance"  # default color for other cases
))
legend_order <- c( "CMD", "UMD", "No significance", "Active site")
#merged_df$key_GnomadV4 <- factor(merged_df$key, levels = legend_order)


# Active site assignment as before execpt with the gnomAD v4 dataframe
active_site_column <- c(40,41,42, 475,476,477)
for (column_number in active_site_column) {
merged_df <- merged_df %>%
  mutate(key_GnomadV4 = case_when(
    column == column_number & key_GnomadV4 == "No significance" ~ "Active site",
    TRUE ~ key_GnomadV4
  )) }



table(merged_df$key_GnomadV4)
merged_df %>% filter(key_GnomadV4 == "CMD")
merged_df %>% filter(key_GnomadV4 == "UMD")
aspartyl_proteases %>% filter(key %in%  c("CMD","UMD", "Active site")) %>% select(column, key)


#column numbers are  2 UMDs and 2 CMDs significant, these are CMD 42 and 753 and UMDs are 299 and 510. 


#table(merged_df$key_GnomadV43) #Put in a bit about showing Active sites are CMDs
```

```{r}
# Splitting the dataframe using active sites and CMDs and UMDs


active_site <- which(merged_df$key == "Active site") 
active_site



CMD_UMD_Active <- merged_df %>%
  filter(key_GnomadV4 %in% c("UMD", "CMD", "Active site"))%>%
  select(column, key_GnomadV4)
CMD_UMD_Active


table(merged_df$key_GnomadV4)  #number of CMDs and UMDs.
```



```{r}

# Plotting the Missense enrichment graph for the gnomAD v4 data 

legend_order <- c( "CMD", "UMD", "No significance",  "Active site")  #Reorder because active site is significant 
merged_df$key_GnomadV4 <- factor(merged_df$key_GnomadV4, levels = legend_order)


your_plot <-ggplot(merged_df, aes(x = shenkin_nrank, y = mes_or, color = key_GnomadV4, alpha = key_GnomadV4)) +
  geom_point() +
  #geom_text(data = aspartyl_proteases, aes(label = column), hjust = -0.2, vjust = 0.5, size = 4) +
  labs(x = "Alignment column conservations  (Shenkin)", y = "Missense score (human variation)") +
  geom_hline(yintercept = 1, linetype = "dashed", color = "black") +
  geom_vline(xintercept = 0.25, linetype = "dashed", color = "black") +
  geom_vline(xintercept = 0.75, linetype = "dashed", color = "black") + 
  #scale_color_manual(values = c("grey" = "grey", "blue" = "blue", "red" = "red", "purple" = "purple", "black"="black" )) + 
  scale_color_manual(values = c("Active site" = "purple", "No significance" = "grey", "black" = "black", "UMD" = "red", "CMD" = "blue", "UME" = "yellow" )) +
  scale_alpha_manual(values = c("Active site" = 0.8, "No significance" = 0.4, "CMD" = 0.8, "UMD" = 0.8)) +
  scale_y_log10()
your_plot


ggsave("C:/Users/Adam Martin/OneDrive - University of Dundee/Documents/Honour Project Dump/MES_shenkin_GnomadV4_Real_MES.png", plot = your_plot, width = 6, height = 4, units = "in")

```

```{r}
# Combines the aspartyl protease with the gnomAD v4 assignments for use in structural analysis.
subset_df <- merged_df[, c("column", "key_GnomadV4")]
aspartyl_proteases <- left_join(aspartyl_proteases, subset_df, by = "column")


#column stops at 770 in merged so as no structural data after this point. aspartyl_protease stops at 788. Loose 18 positions, drop 18 rows

aspartyl_proteases <- na.omit(aspartyl_proteases, cols = "key_GnomadV4")
aspartyl_proteases$key_GnomadV4
```



```{r}
#legend_order <- c( "CMD", "UMD", "Active site", "No significance" )  #Reorder because active site is significant 
#aspartyl_proteases$key_GnomadV4 <- factor(aspartyl_proteases$key_GnomadV4, levels = legend_order)


# Plots the ligand and protein interactions graphs as before, however for the gnomAD v4 data
ligand_or_protein_graph(aspartyl_proteases, aspartyl_proteases$PDB_protein_ligand_interactions,aspartyl_proteases$key_GnomadV4, "ligand interaction",  "C:/Users/Adam Martin/OneDrive - University of Dundee/Documents/Honour Project Dump/ligand_interaction_GnomadV4_Real_MES.png")
ligand_or_protein_graph(aspartyl_proteases, aspartyl_proteases$PDB_protein_protein_interactions,aspartyl_proteases$key_GnomadV4, "protein interaction",  "C:/Users/Adam Martin/OneDrive - University of Dundee/Documents/Honour Project Dump/protein_interaction_GnomadV4_Real_MES.png")





file_path <- "C:/Users/Adam Martin/OneDrive - University of Dundee/Honours Project/231006a-AM-honours-project-aps/varalign-data-PF00026/feather/"
```

```{r}
# Plots the ligand and protein interactions graphs as before in a combined plot using gnomAD v4 data
legend_order <- c( "CMD", "UMD", "Active site", "No significance" )  #Reorder because active site is significant 
aspartyl_proteases$key_GnomadV4 <- factor(aspartyl_proteases$key_GnomadV4, levels = legend_order)


plot1 <- ligand_or_protein_graph(aspartyl_proteases, aspartyl_proteases$PDB_protein_ligand_interactions,aspartyl_proteases$key_GnomadV4, "Ligand",  "C:/Users/Adam Martin/OneDrive - University of Dundee/Documents/Honour Project Dump/ligand_interaction_Real_MES.png")
plot2 <- ligand_or_protein_graph(aspartyl_proteases, aspartyl_proteases$PDB_protein_protein_interactions,aspartyl_proteases$key_GnomadV4, "Protein",  "C:/Users/Adam Martin/OneDrive - University of Dundee/Documents/Honour Project Dump/protein_interaction_Real_MES.png")



plot_list <- list(plot1, plot2)

combined_plot <- plot_grid(
  plotlist = plot_list,
  ncol = 1,
  widths = unit(30, "inches"),
  rel_heights = c(2, 1.4)
)


save_path <- "C:/Users/Adam Martin/OneDrive - University of Dundee/Documents/Honour Project Dump/ligand_protein_interactionGnomadV4_Real_MES.png"
print(combined_plot)
ggsave(save_path, plot = combined_plot, width = 30, height = 15, units = "in")
```


```{r}
# As before using the amino acid postion and secondary structure graphs using the gnomAD v4 data. And the combined graphs used in the supplementary section of thesis
create_custom_plot <- function(selected_columns, index_of_UMD_CMD_active_site, title, key, i) {
  selected_columns$index <- seq_len(nrow(selected_columns))
  df_long <- gather(selected_columns, key = "Variable", value = "Value", -index)
  df_combined <- cbind(df_long, keys = key)
  
  p <- ggplot(df_combined, aes(x = index, y = Value, fill = Variable, alpha = as.factor(index %in% index_of_UMD_CMD_active_site))) +
    geom_bar(stat = "identity") +
    #theme_minimal() +
    labs(x = ifelse(i == "Active site", "Alignment Column", ""), y = i, fill = "") +
    coord_cartesian(ylim = c(0, 16)) +
    scale_y_continuous(breaks = seq(0, 16, by = 4)) +
    scale_x_continuous(breaks = seq(0, nrow(aspartyl_proteases), by = 25)) +
    scale_alpha_manual(values = c("TRUE" = 5, "FALSE" = 0.2), guide = "none") +
    theme(legend.position = ifelse(i == "Active site", "bottom", "none"), legend.justification ="center", plot.margin = margin(0, 0, 0, 0),      plot.title = element_text(hjust = 0.5))
  
  
    if (i == "CMD") {
    p <- p + ggtitle(title)  # Add your desired title
  }

  print(p)
  return(p)
}

read_feather_file <- function(file_path) {
  data <- arrow::read_feather(file_path)
  return(data)
}

create_plot <- function(selected_columns, CMD_UMD_Active, title, key) {
  plot_list <- list()

  for (i in c("CMD", "UMD", "Active site")) {
    index_of_UMD_CMD_active_site <- CMD_UMD_Active %>% filter(key_GnomadV4 == i)
    plot <- create_custom_plot(selected_columns, index_of_UMD_CMD_active_site$dex,  title, key, i)
    plot_list[[i]] <- plot
  }

  combined_plot <- plot_grid(plotlist = plot_list, ncol = 1,  widths = c(8.27), rel_heights = c(1.6, 1.3, 1))
  print(combined_plot)
  return(combined_plot)
}







CMD_UMD_Active <- aspartyl_proteases %>%
  filter(key_GnomadV4 %in% c("UMD", "CMD", "Active site"))%>%
  select(dex, key_GnomadV4)

table(CMD_UMD_Active$key_GnomadV4)


file_path <- "C:/Users/Adam Martin/OneDrive - University of Dundee/Honours Project/231006a-AM-honours-project-aps/varalign-data-PF00026/feather/"

rsa_class <- read_feather_file(paste0(file_path, "rsa_class_df.feather"))
rsa_class <- rsa_class %>% rename(column = "__index_level_0__")
rsa_class <- left_join(aspartyl_proteases, rsa_class, by = "column")
selected_columns <- rsa_class[c("core", "part", "surf")]
plot<-create_plot(selected_columns, CMD_UMD_Active, "Amino acid position", "RSA")
save_path <-"C:/Users/Adam Martin/OneDrive - University of Dundee/Documents/Honour Project Dump/RSA_GnomadV4_Real_MES.png"
ggsave(save_path, plot = plot, width = 6, height = 4, units = "in")

ss_class <- read_feather_file(paste0(file_path,"ss_class_df.feather"))

ss_class <- ss_class %>% rename(column = "__index_level_0__")
ss_class <- left_join(aspartyl_proteases, ss_class, by = "column")
selected_columns <- ss_class[c("helix", "coil", "strand")]
plot<-create_plot(selected_columns, CMD_UMD_Active, "Secondary Structure", "SS")
save_path <-"C:/Users/Adam Martin/OneDrive - University of Dundee/Documents/Honour Project Dump/SS_GnomadV4_Real_MES.png"
ggsave(save_path, plot = plot, width = 6, height = 4, units = "in")



aspartyl_proteases



```


```{r}
# Analysis of the CMDs and UMDs within the gnomAD v4 data
# 2 UMDs and 2 CMDs significant, these are CMD 22 and 307 and UMDs are 199 and 230. 
# CMD 22 is part and coil and high in protein and ligand, CMD 307 is core and strand. 
# UMD 119 high in protein and ligand and 230 high in protein are surface are both coils 



# Looking at the amino acid postion of the CMDs and UMDs
significant <- c("CMD", "UMD")
columns_to_check <- c("p_core", "p_part", "p_surf")
rsa_df <- rsa_class %>%
  filter(key_GnomadV4 %in% significant)%>%
  select("column","dex", "key_GnomadV4", "p_core", "p_part", "p_surf")
rsa_df$max_column <- names(rsa_df[columns_to_check])[max.col(rsa_df[columns_to_check], "first")]
colnames(rsa_df)[colnames(rsa_df) == "max_column"] <- "RSA_max_column"
rsa_df



# Looking at the secondary structure of the CMDs and UMDs
columns_to_check <- c("p_coil", "p_strand", "p_helix")
ss_df <- ss_class %>%
  filter(key_GnomadV4 %in% significant)%>%
  select("column", "key_GnomadV4", "p_coil", "p_strand", "p_helix")
ss_df$max_column <- names(ss_df[columns_to_check])[max.col(ss_df[columns_to_check], "first")]
colnames(ss_df)[colnames(ss_df) == "max_column"] <- "SS_max_column"
ss_df




# Protein ligand interactions and third quartile ranges .CMDs and UMDs
third_quartile <- quantile(aspartyl_proteases$PDB_protein_ligand_interactions, 0.75)
average <- mean(aspartyl_proteases$PDB_protein_ligand_interactions)
third_quartile
average

ligand_df <- aspartyl_proteases %>%
  filter(key_GnomadV4 %in% significant, PDB_protein_ligand_interactions >  third_quartile)%>%
  select( "dex", "column", "key_GnomadV4","PDB_protein_ligand_interactions")
ligand_df


third_quartile <- quantile(aspartyl_proteases$PDB_protein_protein_interactions, 0.75)
average <- mean(aspartyl_proteases$PDB_protein_protein_interactions)
third_quartile
average

protein_df <- aspartyl_proteases %>%
  filter(key_GnomadV4 %in% significant, PDB_protein_protein_interactions >  third_quartile)%>%
  select( "dex", "column", "key_GnomadV4","PDB_protein_protein_interactions")
protein_df



```


```{r}
library(seqinr)

# Used to look at the amino acids within the columns of the CMDs and UMD. 
# Its a function used to calculate the highest number of amino acid within a column and the lowest 

analyze_sinlge_alignment <- function(fasta_file_path) {
  # Read the FASTA file
  fasta_data <- read.fasta(fasta_file_path)

  # Create an empty data frame to store the results
  result_df <- data.frame(sequence_id = character(),
                          amino_acids = character(),
                          stringsAsFactors = FALSE)

  # Iterate through each sequence and extract amino acids from column 1
  for (i in seq_along(fasta_data)) {
    sequence_id <- names(fasta_data)[i]
    amino_acids <- as.character(fasta_data[[i]])

    # Append the results to the data frame
    result_df <- rbind(result_df, data.frame(sequence_id, amino_acids))
  }

  # Combine all amino acids into a single string
  all_amino_acids <- paste(result_df$amino_acids, collapse = "")

  # Function to calculate percentage
  calculate_percentage <- function(amino_acids) {
    counts <- table(strsplit(amino_acids, NULL)[[1]])
    percentages <- prop.table(counts) * 100
    return(data.frame(amino_acid = names(percentages), percentage = as.numeric(percentages)))
  }

  # Calculate percentages for the entire column
  column_percentage <- calculate_percentage(all_amino_acids)

  # Find the row with the maximum and minimum percentages
  max_row <- column_percentage[which.max(column_percentage$percentage), ]
  min_row <- column_percentage[which.min(column_percentage$percentage), ]

  # Calculate the sum of percentages
  total_percentage <- sum(column_percentage$percentage)

  # Return the results as a list
  results <- list(result_df = result_df,
                  column_percentage = column_percentage,
                  max_row = max_row,
                  min_row = min_row,
                  total_percentage = total_percentage)

  return(results)
}







```

```{r}
# Generated a fasta file for each of the columns of interest

# Ran them through the function above to workout the highest and lowest amino acid percentage within the columns


fasta_file <- "C:/Users/Adam Martin/OneDrive - University of Dundee/Documents/Honour Project Dump/Alignment/UMD299.fasta"
results <- analyze_sinlge_alignment(fasta_file)

# Access the results using results$result_df, results$column_percentage, etc.
print(results$result_df)
print(results$column_percentage)
print(results$max_row)
print(results$min_row)
print(results$total_percentage)

#  Eukaryotic  UMD299 is 29% v and 0.7% C

fasta_file <- "C:/Users/Adam Martin/OneDrive - University of Dundee/Documents/Honour Project Dump/Alignment/HumanUMD299.fasta"
results <- analyze_sinlge_alignment(fasta_file)

# Access the results using results$result_df, results$column_percentage, etc.
print(results$result_df)
print(results$column_percentage)
print(results$max_row)
print(results$min_row)
print(results$total_percentage)

#  Human only UMD299 is 40% v and 10% I
```


```{r}
fasta_file <- "C:/Users/Adam Martin/OneDrive - University of Dundee/Documents/Honour Project Dump/Alignment/UMD421.fasta"
results <- analyze_sinlge_alignment(fasta_file)

# Access the results using results$result_df, results$column_percentage, etc.
print(results$result_df)
print(results$column_percentage)
print(results$max_row)
print(results$min_row)
print(results$total_percentage)

#  Eukaryotic  UMD421 is 24.8% S and 0.7% f

fasta_file <- "C:/Users/Adam Martin/OneDrive - University of Dundee/Documents/Honour Project Dump/Alignment/HumanUMD421.fasta"
results <- analyze_sinlge_alignment(fasta_file)

# Access the results using results$result_df, results$column_percentage, etc.
print(results$result_df)
print(results$column_percentage)
print(results$max_row)
print(results$min_row)
print(results$total_percentage)

#  Human only UMD421 is 30% e and 10% k



```
```{r}
fasta_file <- "C:/Users/Adam Martin/OneDrive - University of Dundee/Documents/Honour Project Dump/Alignment/UMD421.fasta"
results <- analyze_sinlge_alignment(fasta_file)

# Access the results using results$result_df, results$column_percentage, etc.
print(results$result_df)
print(results$column_percentage)
print(results$max_row)
print(results$min_row)
print(results$total_percentage)

#  Eukaryotic  UMD421 is 24.8% S and 0.7% f

fasta_file <- "C:/Users/Adam Martin/OneDrive - University of Dundee/Documents/Honour Project Dump/Alignment/HumanUMD421.fasta"
results <- analyze_sinlge_alignment(fasta_file)

# Access the results using results$result_df, results$column_percentage, etc.
print(results$result_df)
print(results$column_percentage)
print(results$max_row)
print(results$min_row)
print(results$total_percentage)

#  Human only UMD421 is 30% e and 10% k

```



```{r}
fasta_file <- "C:/Users/Adam Martin/OneDrive - University of Dundee/Documents/Honour Project Dump/Alignment/UMDgnomADv4_510.fasta"
results <- analyze_sinlge_alignment(fasta_file)

# Access the results using results$result_df, results$column_percentage, etc.
print(results$result_df)
print(results$column_percentage)
print(results$max_row)
print(results$min_row)
print(results$total_percentage)

#  Eukaryotic  UMD230 is 16.8% s and 0.7% f

fasta_file <- "C:/Users/Adam Martin/OneDrive - University of Dundee/Documents/Honour Project Dump/Alignment/HumanUMDgnomADv4_510.fasta"
results <- analyze_sinlge_alignment(fasta_file)

# Access the results using results$result_df, results$column_percentage, etc.
print(results$result_df)
print(results$column_percentage)
print(results$max_row)
print(results$min_row)
print(results$total_percentage)

#  Human only UMD230 is 30% i and 10% d


# Looking at the paalogs of UMD position 230
aspartyl_proteases %>% filter(dex == 230) %>% select(column)
UMD230 <- structure %>% filter(column== 510) 
UMD230$SOURCE_ID_A
unique(UMD230)
names(structure)
```


```{r}
selected_columns <- aspartyl_proteases[, c("column", "dex")]

# Specify the file path where you want to save the CSV file
file_path <- "C:/Users/Adam Martin/OneDrive - University of Dundee/Documents/Honour Project Dump/Alignment/column_to_index.csv"

# Write the selected columns to a CSV file
write.csv(selected_columns, file = file_path, row.names = FALSE)

```




