{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# STRUCTURAL ANALYSIS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Libraries required for the code\n",
    "\n",
    "import pandas as pd\n",
    "import statistics\n",
    "import importlib\n",
    "from scipy import stats\n",
    "import math\n",
    "import numpy as np\n",
    "from pathlib import Path  \n",
    "import vcf\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pyarrow\n",
    "import Bio.SeqIO\n",
    "import Bio.AlignIO\n",
    "from Bio.PDB import *\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reads in pickle file from VarAlign output, containing all structural data\n",
    "Structure = pd.read_pickle(r'C:\\Users\\Adam Martin\\OneDrive - University of Dundee\\Honours Project\\231006a-AM-honours-project-aps\\varalign-data-PF00026\\pickles\\PF00026.22_swissprot.sto_prointvar_structure_table.p.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ATOM_A</th>\n",
       "      <th>ATOM_B</th>\n",
       "      <th>Alignment_column_A</th>\n",
       "      <th>Alignment_column_B</th>\n",
       "      <th>B_iso_or_equiv_A</th>\n",
       "      <th>B_iso_or_equiv_B</th>\n",
       "      <th>CATH_dbAccessionId_A</th>\n",
       "      <th>CATH_dbAccessionId_B</th>\n",
       "      <th>CATH_regionEnd_A</th>\n",
       "      <th>CATH_regionEnd_B</th>\n",
       "      <th>...</th>\n",
       "      <th>pdbx_formal_charge_B</th>\n",
       "      <th>type_symbol_A</th>\n",
       "      <th>type_symbol_B</th>\n",
       "      <th>interaction_type</th>\n",
       "      <th>protein_topology</th>\n",
       "      <th>pfam_domain_topology</th>\n",
       "      <th>cath_domain_topology</th>\n",
       "      <th>cath_hierarchical_topology</th>\n",
       "      <th>scop_domain_topology</th>\n",
       "      <th>polymer_topology</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C</td>\n",
       "      <td>N</td>\n",
       "      <td>230.0</td>\n",
       "      <td>232.0</td>\n",
       "      <td>16.675000</td>\n",
       "      <td>12.560000</td>\n",
       "      <td>2.40.70.10</td>\n",
       "      <td>2.40.70.10</td>\n",
       "      <td>214.0</td>\n",
       "      <td>214.0</td>\n",
       "      <td>...</td>\n",
       "      <td>?</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>Protein-Protein</td>\n",
       "      <td>Homoprotein</td>\n",
       "      <td>Homodomain (Self)</td>\n",
       "      <td>Homodomain</td>\n",
       "      <td>level_4</td>\n",
       "      <td>Homodomain</td>\n",
       "      <td>Intrapolymer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>O</td>\n",
       "      <td>N</td>\n",
       "      <td>177.0</td>\n",
       "      <td>232.0</td>\n",
       "      <td>34.367500</td>\n",
       "      <td>12.560000</td>\n",
       "      <td>2.40.70.10</td>\n",
       "      <td>2.40.70.10</td>\n",
       "      <td>214.0</td>\n",
       "      <td>214.0</td>\n",
       "      <td>...</td>\n",
       "      <td>?</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>Protein-Protein</td>\n",
       "      <td>Homoprotein</td>\n",
       "      <td>Homodomain (Self)</td>\n",
       "      <td>Homodomain</td>\n",
       "      <td>level_4</td>\n",
       "      <td>Homodomain</td>\n",
       "      <td>Intrapolymer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>O</td>\n",
       "      <td>O</td>\n",
       "      <td>230.0</td>\n",
       "      <td>311.0</td>\n",
       "      <td>16.675000</td>\n",
       "      <td>18.305714</td>\n",
       "      <td>2.40.70.10</td>\n",
       "      <td>2.40.70.10</td>\n",
       "      <td>214.0</td>\n",
       "      <td>214.0</td>\n",
       "      <td>...</td>\n",
       "      <td>?</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>Protein-Protein</td>\n",
       "      <td>Homoprotein</td>\n",
       "      <td>Homodomain (Self)</td>\n",
       "      <td>Homodomain</td>\n",
       "      <td>level_4</td>\n",
       "      <td>Homodomain</td>\n",
       "      <td>Intrapolymer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>O</td>\n",
       "      <td>N</td>\n",
       "      <td>300.0</td>\n",
       "      <td>311.0</td>\n",
       "      <td>47.157500</td>\n",
       "      <td>18.305714</td>\n",
       "      <td>2.40.70.10</td>\n",
       "      <td>2.40.70.10</td>\n",
       "      <td>214.0</td>\n",
       "      <td>214.0</td>\n",
       "      <td>...</td>\n",
       "      <td>?</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>Protein-Protein</td>\n",
       "      <td>Homoprotein</td>\n",
       "      <td>Homodomain (Self)</td>\n",
       "      <td>Homodomain</td>\n",
       "      <td>level_4</td>\n",
       "      <td>Homodomain</td>\n",
       "      <td>Intrapolymer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NE2</td>\n",
       "      <td>O</td>\n",
       "      <td>229.0</td>\n",
       "      <td>311.0</td>\n",
       "      <td>16.737778</td>\n",
       "      <td>18.305714</td>\n",
       "      <td>2.40.70.10</td>\n",
       "      <td>2.40.70.10</td>\n",
       "      <td>214.0</td>\n",
       "      <td>214.0</td>\n",
       "      <td>...</td>\n",
       "      <td>?</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>Protein-Protein</td>\n",
       "      <td>Homoprotein</td>\n",
       "      <td>Homodomain (Self)</td>\n",
       "      <td>Homodomain</td>\n",
       "      <td>level_4</td>\n",
       "      <td>Homodomain</td>\n",
       "      <td>Intrapolymer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2222791</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>424.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.920000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.40.70.10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>383.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain-NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interpolymer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2222792</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>120.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>62.100000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.40.70.10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>147.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain-NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interpolymer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2222793</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>144.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>56.217778</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.40.70.10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>147.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain-NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interpolymer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2222794</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>227.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>43.244286</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.40.70.10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>147.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain-NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interpolymer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2222795</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>331.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>43.510000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.40.70.10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>147.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain-NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Interpolymer</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2222796 rows Ã— 166 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ATOM_A ATOM_B  Alignment_column_A  Alignment_column_B  \\\n",
       "0            C      N               230.0               232.0   \n",
       "1            O      N               177.0               232.0   \n",
       "2            O      O               230.0               311.0   \n",
       "3            O      N               300.0               311.0   \n",
       "6          NE2      O               229.0               311.0   \n",
       "...        ...    ...                 ...                 ...   \n",
       "2222791    NaN    NaN               424.0                 NaN   \n",
       "2222792    NaN    NaN               120.0                 NaN   \n",
       "2222793    NaN    NaN               144.0                 NaN   \n",
       "2222794    NaN    NaN               227.0                 NaN   \n",
       "2222795    NaN    NaN               331.0                 NaN   \n",
       "\n",
       "         B_iso_or_equiv_A  B_iso_or_equiv_B CATH_dbAccessionId_A  \\\n",
       "0               16.675000         12.560000           2.40.70.10   \n",
       "1               34.367500         12.560000           2.40.70.10   \n",
       "2               16.675000         18.305714           2.40.70.10   \n",
       "3               47.157500         18.305714           2.40.70.10   \n",
       "6               16.737778         18.305714           2.40.70.10   \n",
       "...                   ...               ...                  ...   \n",
       "2222791         41.920000               NaN           2.40.70.10   \n",
       "2222792         62.100000               NaN           2.40.70.10   \n",
       "2222793         56.217778               NaN           2.40.70.10   \n",
       "2222794         43.244286               NaN           2.40.70.10   \n",
       "2222795         43.510000               NaN           2.40.70.10   \n",
       "\n",
       "        CATH_dbAccessionId_B  CATH_regionEnd_A  CATH_regionEnd_B  ...  \\\n",
       "0                 2.40.70.10             214.0             214.0  ...   \n",
       "1                 2.40.70.10             214.0             214.0  ...   \n",
       "2                 2.40.70.10             214.0             214.0  ...   \n",
       "3                 2.40.70.10             214.0             214.0  ...   \n",
       "6                 2.40.70.10             214.0             214.0  ...   \n",
       "...                      ...               ...               ...  ...   \n",
       "2222791                  NaN             383.0               NaN  ...   \n",
       "2222792                  NaN             147.0               NaN  ...   \n",
       "2222793                  NaN             147.0               NaN  ...   \n",
       "2222794                  NaN             147.0               NaN  ...   \n",
       "2222795                  NaN             147.0               NaN  ...   \n",
       "\n",
       "        pdbx_formal_charge_B type_symbol_A  type_symbol_B  interaction_type  \\\n",
       "0                          ?             N              N   Protein-Protein   \n",
       "1                          ?             N              N   Protein-Protein   \n",
       "2                          ?             N              N   Protein-Protein   \n",
       "3                          ?             N              N   Protein-Protein   \n",
       "6                          ?             N              N   Protein-Protein   \n",
       "...                      ...           ...            ...               ...   \n",
       "2222791                  NaN             N            NaN               NaN   \n",
       "2222792                  NaN             N            NaN               NaN   \n",
       "2222793                  NaN             N            NaN               NaN   \n",
       "2222794                  NaN             N            NaN               NaN   \n",
       "2222795                  NaN             N            NaN               NaN   \n",
       "\n",
       "         protein_topology  pfam_domain_topology cath_domain_topology  \\\n",
       "0             Homoprotein     Homodomain (Self)           Homodomain   \n",
       "1             Homoprotein     Homodomain (Self)           Homodomain   \n",
       "2             Homoprotein     Homodomain (Self)           Homodomain   \n",
       "3             Homoprotein     Homodomain (Self)           Homodomain   \n",
       "6             Homoprotein     Homodomain (Self)           Homodomain   \n",
       "...                   ...                   ...                  ...   \n",
       "2222791               NaN                   NaN           Domain-NaN   \n",
       "2222792               NaN                   NaN           Domain-NaN   \n",
       "2222793               NaN                   NaN           Domain-NaN   \n",
       "2222794               NaN                   NaN           Domain-NaN   \n",
       "2222795               NaN                   NaN           Domain-NaN   \n",
       "\n",
       "        cath_hierarchical_topology  scop_domain_topology  polymer_topology  \n",
       "0                          level_4            Homodomain      Intrapolymer  \n",
       "1                          level_4            Homodomain      Intrapolymer  \n",
       "2                          level_4            Homodomain      Intrapolymer  \n",
       "3                          level_4            Homodomain      Intrapolymer  \n",
       "6                          level_4            Homodomain      Intrapolymer  \n",
       "...                            ...                   ...               ...  \n",
       "2222791                        NaN                   NaN      Interpolymer  \n",
       "2222792                        NaN                   NaN      Interpolymer  \n",
       "2222793                        NaN                   NaN      Interpolymer  \n",
       "2222794                        NaN                   NaN      Interpolymer  \n",
       "2222795                        NaN                   NaN      Interpolymer  \n",
       "\n",
       "[2222796 rows x 166 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Viewing structure table\n",
    "Structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ATOM_A',\n",
       " 'ATOM_B',\n",
       " 'Alignment_column_A',\n",
       " 'Alignment_column_B',\n",
       " 'B_iso_or_equiv_A',\n",
       " 'B_iso_or_equiv_B',\n",
       " 'CATH_dbAccessionId_A',\n",
       " 'CATH_dbAccessionId_B',\n",
       " 'CATH_regionEnd_A',\n",
       " 'CATH_regionEnd_B',\n",
       " 'CATH_regionId_A',\n",
       " 'CATH_regionId_B',\n",
       " 'CATH_regionResNum_A',\n",
       " 'CATH_regionResNum_B',\n",
       " 'CATH_regionStart_A',\n",
       " 'CATH_regionStart_B',\n",
       " 'CHAIN_A',\n",
       " 'CHAIN_B',\n",
       " 'Cartn_x_A',\n",
       " 'Cartn_x_B',\n",
       " 'Cartn_y_A',\n",
       " 'Cartn_y_B',\n",
       " 'Cartn_z_A',\n",
       " 'Cartn_z_B',\n",
       " 'DIST',\n",
       " 'Ensembl_dbAccessionId_A',\n",
       " 'Ensembl_dbAccessionId_B',\n",
       " 'Ensembl_dbExonId_A',\n",
       " 'Ensembl_dbExonId_B',\n",
       " 'Ensembl_dbTranscriptId_A',\n",
       " 'Ensembl_dbTranscriptId_B',\n",
       " 'Ensembl_dbTranslationId_A',\n",
       " 'Ensembl_dbTranslationId_B',\n",
       " 'Ensembl_regionEnd_A',\n",
       " 'Ensembl_regionEnd_B',\n",
       " 'Ensembl_regionId_A',\n",
       " 'Ensembl_regionId_B',\n",
       " 'Ensembl_regionResNum_A',\n",
       " 'Ensembl_regionResNum_B',\n",
       " 'Ensembl_regionStart_A',\n",
       " 'Ensembl_regionStart_B',\n",
       " 'INSCODE_A',\n",
       " 'INSCODE_B',\n",
       " 'Int_Types',\n",
       " 'PDB_Annotation_A',\n",
       " 'PDB_Annotation_B',\n",
       " 'PDB_codeSecondaryStructure_A',\n",
       " 'PDB_codeSecondaryStructure_B',\n",
       " 'PDB_dbAccessionId_A',\n",
       " 'PDB_dbAccessionId_B',\n",
       " 'PDB_dbChainId_A',\n",
       " 'PDB_dbChainId_B',\n",
       " 'PDB_dbResName_A',\n",
       " 'PDB_dbResName_B',\n",
       " 'PDB_dbResNum_A',\n",
       " 'PDB_dbResNum_B',\n",
       " 'PDB_entityId_A',\n",
       " 'PDB_entityId_B',\n",
       " 'PDB_nameSecondaryStructure_A',\n",
       " 'PDB_nameSecondaryStructure_B',\n",
       " 'PDB_regionEnd_A',\n",
       " 'PDB_regionEnd_B',\n",
       " 'PDB_regionId_A',\n",
       " 'PDB_regionId_B',\n",
       " 'PDB_regionResNum_A',\n",
       " 'PDB_regionResNum_B',\n",
       " 'PDB_regionStart_A',\n",
       " 'PDB_regionStart_B',\n",
       " 'Pfam_dbAccessionId_A',\n",
       " 'Pfam_dbAccessionId_B',\n",
       " 'Pfam_regionEnd_A',\n",
       " 'Pfam_regionEnd_B',\n",
       " 'Pfam_regionId_A',\n",
       " 'Pfam_regionId_B',\n",
       " 'Pfam_regionResNum_A',\n",
       " 'Pfam_regionResNum_B',\n",
       " 'Pfam_regionStart_A',\n",
       " 'Pfam_regionStart_B',\n",
       " 'RES_A',\n",
       " 'RES_B',\n",
       " 'RES_FULL_A',\n",
       " 'RES_FULL_B',\n",
       " 'SCOP_dbAccessionId_A',\n",
       " 'SCOP_dbAccessionId_B',\n",
       " 'SCOP_regionEnd_A',\n",
       " 'SCOP_regionEnd_B',\n",
       " 'SCOP_regionId_A',\n",
       " 'SCOP_regionId_B',\n",
       " 'SCOP_regionResNum_A',\n",
       " 'SCOP_regionResNum_B',\n",
       " 'SCOP_regionStart_A',\n",
       " 'SCOP_regionStart_B',\n",
       " 'SOURCE_ID_A',\n",
       " 'SOURCE_ID_B',\n",
       " 'UniProt_dbAccessionId_A',\n",
       " 'UniProt_dbAccessionId_B',\n",
       " 'UniProt_dbResName_A',\n",
       " 'UniProt_dbResName_B',\n",
       " 'UniProt_dbResNum_A',\n",
       " 'UniProt_dbResNum_B',\n",
       " 'UniProt_regionEnd_A',\n",
       " 'UniProt_regionEnd_B',\n",
       " 'UniProt_regionId_A',\n",
       " 'UniProt_regionId_B',\n",
       " 'UniProt_regionResNum_A',\n",
       " 'UniProt_regionResNum_B',\n",
       " 'UniProt_regionStart_A',\n",
       " 'UniProt_regionStart_B',\n",
       " 'VDW_DIST',\n",
       " 'auth_asym_id_A',\n",
       " 'auth_asym_id_B',\n",
       " 'auth_atom_altloc_id_A',\n",
       " 'auth_atom_altloc_id_B',\n",
       " 'auth_atom_id_A',\n",
       " 'auth_atom_id_B',\n",
       " 'auth_comp_id_A',\n",
       " 'auth_comp_id_B',\n",
       " 'auth_seq_id_A',\n",
       " 'auth_seq_id_B',\n",
       " 'auth_seq_id_full_A',\n",
       " 'auth_seq_id_full_B',\n",
       " 'group_PDB_A',\n",
       " 'group_PDB_B',\n",
       " 'id_A',\n",
       " 'id_B',\n",
       " 'index_A',\n",
       " 'index_B',\n",
       " 'label_alt_id_A',\n",
       " 'label_alt_id_B',\n",
       " 'label_asym_id_A',\n",
       " 'label_asym_id_B',\n",
       " 'label_atom_altloc_id_A',\n",
       " 'label_atom_altloc_id_B',\n",
       " 'label_atom_id_A',\n",
       " 'label_atom_id_B',\n",
       " 'label_comp_id_A',\n",
       " 'label_comp_id_B',\n",
       " 'label_entity_id_A',\n",
       " 'label_entity_id_B',\n",
       " 'label_seq_id_A',\n",
       " 'label_seq_id_B',\n",
       " 'label_seq_id_full_A',\n",
       " 'label_seq_id_full_B',\n",
       " 'new_asym_id_A',\n",
       " 'new_asym_id_B',\n",
       " 'new_seq_id_A',\n",
       " 'new_seq_id_B',\n",
       " 'occupancy_A',\n",
       " 'occupancy_B',\n",
       " 'pdbe_label_seq_id_A',\n",
       " 'pdbe_label_seq_id_B',\n",
       " 'pdbx_PDB_ins_code_A',\n",
       " 'pdbx_PDB_ins_code_B',\n",
       " 'pdbx_PDB_model_num_A',\n",
       " 'pdbx_PDB_model_num_B',\n",
       " 'pdbx_formal_charge_A',\n",
       " 'pdbx_formal_charge_B',\n",
       " 'type_symbol_A',\n",
       " 'type_symbol_B',\n",
       " 'interaction_type',\n",
       " 'protein_topology',\n",
       " 'pfam_domain_topology',\n",
       " 'cath_domain_topology',\n",
       " 'cath_hierarchical_topology',\n",
       " 'scop_domain_topology',\n",
       " 'polymer_topology']"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# All the different columns containing lots of information\n",
    "Structure.columns.to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reads in pickle file from VarAlign output, containing all DSSP data\n",
    "dssp = pd.read_pickle('C:\\\\Users\\\\Adam Martin\\\\OneDrive - University of Dundee\\\\Honours Project\\\\231006a-AM-honours-project-aps\\\\varalign-data-PF00026\\\\pickles\\\\mapped_dssp_table.p.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This combines the Structure and dssp table, on the protein accession abd protein chain.\n",
    "merged = pd.merge(Structure, dssp, left_on=['PDB_dbAccessionId_A', 'PDB_dbChainId_A', 'PDB_dbResNum_A'],\n",
    "                  right_on=['PDB_dbAccessionId', 'PDB_dbChainId', 'PDB_dbResNum'],\n",
    "                  how='left')#, validate='m:1')\n",
    "\n",
    "\n",
    "# Renames some columns for later analysis\n",
    "merged.rename(columns={'SS': 'SS_A', 'SS_CLASS' : 'SS_CLASS_A', 'SS_CLASS_UNB': 'SS_CLASS_UNB_A',  'RSA': 'RSA_A', 'RSA_CLASS': 'RSA_CLASS_A', 'RSA_UNB' :'RSA_UNB_A', 'RSA_CLASS_UNB' : 'RSA_CLASS_UNB_A', }, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fill in the gaps within the secondary structure assingment\n",
    "merged.SS_A = merged.SS_A.fillna(\"\") # FILL NA WITH \"\" SO WE CAN PROCESS THE SECONDARY STRUCTURE DATA. NA ARE FOUND ON THOSE RESIDUES PRESENTING NO SS, i.e., COIL\n",
    "print(merged.SS_A)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Outputs the structure table as a feather file. This is a file format compatabile with python and R. \n",
    "merged.to_feather(\"C:\\\\Users\\\\Adam Martin\\\\OneDrive - University of Dundee\\\\Honours Project\\\\231006a-AM-honours-project-aps\\\\varalign-data-PF00026\\\\feather\\\\PF00026.22_swissprot.sto_prointvar_structure_table.p.gz.feather\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Positions with human residues: [128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 151, 152, 153, 154, 155, 156, 157, 158, 167, 168, 169, 172, 173, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 207, 208, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 232, 233, 234, 235, 236, 237, 238, 239, 240, 241, 255, 256, 257, 258, 259, 260, 261, 262, 263, 265, 266, 267, 268, 269, 270, 271, 272, 273, 274, 275, 276, 277, 279, 280, 281, 282, 283, 284, 285, 286, 287, 288, 289, 290, 291, 292, 293, 294, 295, 296, 297, 298, 299, 300, 301, 302, 303, 304, 305, 306, 307, 308, 309, 310, 311, 312, 313, 314, 315, 316, 317, 318, 319, 320, 321, 386, 390, 391, 392, 393, 394, 395, 396, 397, 398, 399, 400, 401, 402, 403, 404, 405, 406, 407, 410, 411, 412, 413, 414, 415, 416, 417, 418, 419, 420, 421, 422, 423, 425, 426, 427, 428, 429, 430, 431, 432, 433, 434, 435, 436, 437, 438, 439, 440, 441, 442, 443, 444, 445, 446, 447, 448, 449, 451, 452, 453, 481, 482, 483, 484, 485, 486, 487, 488, 489, 496, 497, 498, 499, 500, 501, 502, 503, 504, 506, 507, 508, 509, 510, 511, 512, 513, 514, 515, 516, 517, 518, 519, 520, 521, 522, 523, 524, 525, 538, 539, 540, 541, 542, 543, 544, 545, 546, 547, 548, 549, 550, 552, 553, 554, 555, 556, 557, 558, 559, 560, 561, 562, 563, 564, 565, 566, 567, 568, 569, 570, 571, 572, 573, 574, 582, 583, 584, 585, 586, 587, 588, 589, 590, 591, 592, 593, 594, 595, 596, 597, 598, 599, 600, 601, 602, 603, 604, 605, 606, 607, 608, 609, 610, 611, 612, 613, 631, 632, 633, 634, 636, 637, 638, 639, 640, 641, 642, 643, 644, 645, 646, 647, 648, 649, 650, 651, 652, 653, 679, 680, 681, 682, 683, 684, 685, 686, 687, 688, 689, 690, 691, 692, 693, 694, 695, 696, 697, 698, 699, 700, 701, 702, 703, 704, 705, 706, 719, 720, 721, 723, 726, 727, 728, 729, 730, 731, 732, 837, 838, 839, 840, 841, 842, 843, 844, 845, 846, 847, 848, 849, 850, 851, 852, 853, 854, 855, 856, 857, 858, 859, 860, 861, 865, 866, 867, 868, 869, 870, 871, 872, 873, 874, 875, 876, 877, 878, 879, 880, 881, 882, 883, 885, 886, 887, 888, 889, 890, 891, 892, 893, 894, 895, 896, 897, 898, 899, 900, 909, 910, 911, 912, 913, 914, 915, 916, 917, 918, 919, 920, 921, 922, 923, 924, 925, 926, 927, 928, 929, 930, 931, 933, 934, 935, 936, 937, 938, 939, 940, 941, 942, 943, 944, 945, 946, 947, 948, 949, 950, 951, 952, 956, 957, 958, 959, 960, 961, 962, 963, 964, 965, 966, 967, 968, 969, 970, 971, 972, 973, 974, 975, 976, 977, 978, 979, 980, 981, 982, 983, 989, 990, 991, 992, 993, 997, 998, 1001, 1002, 1003, 1010, 1021, 1022, 1104, 1105, 1106, 1109, 1110, 1111, 1112, 1113, 1114, 1115, 1116, 1117, 1118, 1119, 1120, 1121, 1122, 1123, 1124, 1125, 1126, 1127, 1128, 1139, 1140, 1141, 1142, 1143, 1144, 1145, 1146, 1147, 1148, 1149, 1150, 1151, 1152, 1153, 1154, 1155, 1156, 1157, 1158, 1159, 1160, 1161, 1162, 1163, 1164, 1165, 1166, 1167, 1168, 1169, 1170, 1171, 1172, 1173, 1174, 1175, 1176, 1177, 1178, 1179, 1180, 1181, 1182]\n",
      "Length: 596\n",
      "[128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 151, 152, 153, 154, 155, 156, 157, 158, 167, 168, 169, 172, 173, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 207, 208, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 232, 233, 234, 235, 236, 237, 238, 239, 240, 241, 255, 256, 257, 258, 259, 260, 261, 262, 263, 265, 266, 267, 268, 269, 270, 271, 272, 273, 274, 275, 276, 277, 279, 280, 281, 282, 283, 284, 285, 286, 287, 288, 289, 290, 291, 292, 293, 294, 295, 296, 297, 298, 299, 300, 301, 302, 303, 304, 305, 306, 307, 308, 309, 310, 311, 312, 313, 314, 315, 316, 317, 318, 319, 320, 321, 386, 390, 391, 392, 393, 394, 395, 396, 397, 398, 399, 400, 401, 402, 403, 404, 405, 406, 407, 410, 411, 412, 413, 414, 415, 416, 417, 418, 419, 420, 421, 422, 423, 425, 426, 427, 428, 429, 430, 431, 432, 433, 434, 435, 436, 437, 438, 439, 440, 441, 442, 443, 444, 445, 446, 447, 448, 449, 451, 452, 453, 481, 482, 483, 484, 485, 486, 487, 488, 489, 496, 497, 498, 499, 500, 501, 502, 503, 504, 506, 507, 508, 509, 510, 511, 512, 513, 514, 515, 516, 517, 518, 519, 520, 521, 522, 523, 524, 525, 538, 539, 540, 541, 542, 543, 544, 545, 546, 547, 548, 549, 550, 552, 553, 554, 555, 556, 557, 558, 559, 560, 561, 562, 563, 564, 565, 566, 567, 568, 569, 570, 571, 572, 573, 574, 582, 583, 584, 585, 586, 587, 588, 589, 590, 591, 592, 593, 594, 595, 596, 597, 598, 599, 600, 601, 602, 603, 604, 605, 606, 607, 608, 609, 610, 611, 612, 613, 631, 632, 633, 634, 636, 637, 638, 639, 640, 641, 642, 643, 644, 645, 646, 647, 648, 649, 650, 651, 652, 653, 679, 680, 681, 682, 683, 684, 685, 686, 687, 688, 689, 690, 691, 692, 693, 694, 695, 696, 697, 698, 699, 700, 701, 702, 703, 704, 705, 706, 719, 720, 721, 723, 726, 727, 728, 729, 730, 731, 732, 837, 838, 839, 840, 841, 842, 843, 844, 845, 846, 847, 848, 849, 850, 851, 852, 853, 854, 855, 856, 857, 858, 859, 860, 861, 865, 866, 867, 868, 869, 870, 871, 872, 873, 874, 875, 876, 877, 878, 879, 880, 881, 882, 883, 885, 886, 887, 888, 889, 890, 891, 892, 893, 894, 895, 896, 897, 898, 899, 900, 909, 910, 911, 912, 913, 914, 915, 916, 917, 918, 919, 920, 921, 922, 923, 924, 925, 926, 927, 928, 929, 930, 931, 933, 934, 935, 936, 937, 938, 939, 940, 941, 942, 943, 944, 945, 946, 947, 948, 949, 950, 951, 952, 956, 957, 958, 959, 960, 961, 962, 963, 964, 965, 966, 967, 968, 969, 970, 971, 972, 973, 974, 975, 976, 977, 978, 979, 980, 981, 982, 983, 989, 990, 991, 992, 993, 997, 998, 1001, 1002, 1003, 1010, 1021, 1022, 1104, 1105, 1106, 1109, 1110, 1111, 1112, 1113, 1114, 1115, 1116, 1117, 1118, 1119, 1120, 1121, 1122, 1123, 1124, 1125, 1126, 1127, 1128, 1139, 1140, 1141, 1142, 1143, 1144, 1145, 1146, 1147, 1148, 1149, 1150, 1151, 1152, 1153, 1154, 1155, 1156, 1157, 1158, 1159, 1160, 1161, 1162, 1163, 1164, 1165, 1166, 1167, 1168, 1169, 1170, 1171, 1172, 1173, 1174, 1175, 1176, 1177, 1178, 1179, 1180, 1181, 1182]\n"
     ]
    }
   ],
   "source": [
    "# This calculates the consequence sequences for the alignemnt. Selecting only columns with human sequence occupancy > 50% \n",
    "def get_cons_cols(aln_in, fmt_in, t = 0.50): #returns the columns with a relative occupancy greater than a threshold\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    from Bio import AlignIO\n",
    "    # Read the alignment\n",
    "    alignment = AlignIO.read(aln_in, 'stockholm')\n",
    "    human_indices = [i for i, seq_record in enumerate(alignment) if 'HUMAN' in seq_record.id]\n",
    "    # Find positions with human residues in the alignment\n",
    "    human_positions = [position + 1 for position in range(alignment.get_alignment_length()) if any(alignment[seq_index, position] != '-' for seq_index in human_indices)]\n",
    "    # Print the positions with human residues\n",
    "    print(\"Positions with human residues:\", human_positions)\n",
    "    print(\"Length:\", len(human_positions))\n",
    "    return human_positions\n",
    "human_cons_cols = get_cons_cols(r'C:\\Users\\Adam Martin\\OneDrive - University of Dundee\\Honours Project\\Jalview and Chimera using the alignment from PF00026 first from linux\\Uniprot_Sequences_For_Alignment_To_Allow_VarAlign_TO\\Uniprot_Annotated_Alignment127_PFAM.sto', 'stockholm') # GETS CONSENSUS COLUMNS FROM MSA (OCCUPANCY > 0.5)\n",
    "print(human_cons_cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Code taken from https://github.com/bartongroup/JSU_ANK_analysis.\n",
    "# Function that works out the concensus relative solvent accessability (RSA) for each of the 16 proteins, it is then combined together\n",
    "def get_rsa_class_consensus(df, aln_cols):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    df_unique = df.drop_duplicates(['UniProt_dbAccessionId_A', 'PDB_dbAccessionId_A', 'PDB_dbChainId_A', 'PDB_dbResNum_A'])\n",
    "    df_unique = df_unique.dropna(subset = [\"RSA_CLASS_A\", \"RSA_A\"])            \n",
    "    df_unique = df_unique.query('RSA_A != \"\"')\n",
    "    df_unique.Alignment_column_A = df_unique.Alignment_column_A.astype(float).astype(int)\n",
    "    df_cons_cols = df_unique[df_unique.Alignment_column_A.isin(aln_cols)]\n",
    "\n",
    "    grouped_col_rep = df_cons_cols.groupby([\"Alignment_column_A\", \"SOURCE_ID_A\"])\n",
    "    rsa_class_dict = {}\n",
    "    rsa_dict = {}\n",
    "    i = 1\n",
    "    for k, v in grouped_col_rep:\n",
    "        col = k[0]\n",
    "        if col not in rsa_class_dict:\n",
    "            print(\"Processing consensus column {}\".format(i))\n",
    "            i += 1\n",
    "            rsa_class_dict[col] = []\n",
    "        if col not in rsa_dict:\n",
    "            rsa_dict[col] = []\n",
    "        \n",
    "        rsa_list = list(map(float, v.RSA_A.tolist()))            \n",
    "        rsa_class_list = v.RSA_CLASS_A.tolist()                 \n",
    "        try:\n",
    "            rsa_class_cons = statistics.mode(rsa_class_list)\n",
    "        except:\n",
    "            rsa_class_cons = rsa_class_list[0]\n",
    "        rsa_median = statistics.median(rsa_list)\n",
    "        rsa_class_dict[col].append(rsa_class_cons)\n",
    "        rsa_dict[col].append(rsa_median)\n",
    "    return rsa_class_dict, rsa_dict\n",
    "\n",
    "\n",
    "rsa_class_dict, rsa_dict = get_rsa_class_consensus(merged, human_cons_cols) # RETRIEVES RSA DATA FROM TABLE\n",
    "print(len(rsa_class_dict))\n",
    "print(rsa_class_dict)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Code taken from https://github.com/bartongroup/JSU_ANK_analysis.\n",
    "# Function that works out the relative solvent assessability assingment proportion for all the proteins\n",
    "def get_rsa_class_df(rsa_class_dict, cons_cols_allsp):\n",
    "    n = len(rsa_class_dict.keys())\n",
    "    core = []\n",
    "    surf = []\n",
    "    part = []\n",
    "    for v in rsa_class_dict.values():\n",
    "        core.append(v.count(\"Core\"))\n",
    "        surf.append(v.count(\"Surface\"))\n",
    "        part.append(v.count(\"Part. Exposed\")) \n",
    "    df_dssp = pd.DataFrame(list(zip(core, surf, part)), columns =  [\"core\", \"surf\", \"part\"])\n",
    "    df_dssp[\"tot\"] = df_dssp.core + df_dssp.surf + df_dssp.part\n",
    "    df_dssp[\"p_core\"] = df_dssp.core / df_dssp.tot\n",
    "    df_dssp[\"p_surf\"] = df_dssp.surf / df_dssp.tot\n",
    "    df_dssp[\"p_part\"] = df_dssp.part / df_dssp.tot\n",
    "    df_dssp[\"p_tot\"] = df_dssp.p_core + df_dssp.p_surf + df_dssp.p_part\n",
    "    df_dssp.index = cons_cols_allsp         #\n",
    "    return df_dssp\n",
    "\n",
    "\n",
    "rsa_class_df = get_rsa_class_df(rsa_class_dict, human_cons_cols) # CALCULATES PROPORTION OF RSA CLASS PER POSITION\n",
    "rsa_class_df\n",
    "\n",
    "# Output the RSA class assingments proportions as a feather \n",
    "rsa_class_df.to_feather(\"/homes/2414054/Honours/231006a-AM-honours-project-aps/varalign-data-PF00026/feather/rsa_class_df.feather\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Code taken from https://github.com/bartongroup/JSU_ANK_analysis.\n",
    "# Function that works out the concensus secondary structure for each of the 16 proteins, it is then combined together\n",
    "def get_ss_class_consensus(df, aln_cols):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    df_unique = df.drop_duplicates(['UniProt_dbAccessionId_A', 'PDB_dbAccessionId_A', 'PDB_dbChainId_A', 'PDB_dbResNum_A'])\n",
    "    df_unique = df_unique.dropna(subset = [\"SS_CLASS_A\"])\n",
    "    df_unique.Alignment_column_A = df_unique.Alignment_column_A.astype(float).astype(int)\n",
    "    df_cons_cols = df_unique[df_unique.Alignment_column_A.isin(aln_cols)]\n",
    "    missing = []\n",
    "    for x in aln_cols:\n",
    "        if x in df_cons_cols['Alignment_column_A'].values:\n",
    "            t = 0\n",
    "        else: \n",
    "            missing.append(x)\n",
    "\n",
    "    grouped_col_rep = df_cons_cols.groupby([\"Alignment_column_A\", \"SOURCE_ID_A\"])\n",
    "    ss_class_dict = {}\n",
    "    i = 1\n",
    "    for k, v in grouped_col_rep:\n",
    "        col = k[0]\n",
    "        if col not in ss_class_dict:\n",
    "\n",
    "            i += 1\n",
    "            ss_class_dict[col] = []\n",
    "        \n",
    "        for x in missing:\n",
    "            ss_class_dict[x] = []\n",
    "            \n",
    "        \n",
    "        ss_class_list = v.SS_CLASS_A.tolist()\n",
    "\n",
    "        try:\n",
    "            ss_class_cons = statistics.mode(ss_class_list)\n",
    "        except:\n",
    "            ss_class_cons = ss_class_list[0]\n",
    "        ss_class_dict[col].append(ss_class_cons)\n",
    "        ss_class_dict = dict(sorted(ss_class_dict.items()))\n",
    "    return ss_class_dict\n",
    "ss_class_dict = get_ss_class_consensus(merged, human_cons_cols) # EXTRACTS SS DATA FROM TABLE AND RETURNS DICTIONARY WITH OCCURRENCE OF SS CLASS PER POSTITION\n",
    "print(len(ss_class_dict))\n",
    "print(ss_class_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Code taken from https://github.com/bartongroup/JSU_ANK_analysis.\n",
    "# Function that works out the secondary structure assingment proportion for all the proteins\n",
    "def get_ss_class_df(rsa_class_dict):\n",
    "    n = len(rsa_class_dict.keys())\n",
    "    h = []\n",
    "    c = []\n",
    "    e = []\n",
    "    for v in rsa_class_dict.values():\n",
    "        h.append(v.count(\"H\"))\n",
    "        c.append(v.count(\"C\"))\n",
    "        e.append(v.count(\"E\")) \n",
    "    df_dssp = pd.DataFrame(list(zip(h, c, e)), columns =  [\"helix\", \"coil\", \"strand\"])\n",
    "    df_dssp[\"tot\"] = df_dssp.helix + df_dssp.coil + df_dssp.strand\n",
    "    df_dssp[\"p_helix\"] = df_dssp.helix / df_dssp.tot\n",
    "    df_dssp[\"p_coil\"] = df_dssp.coil / df_dssp.tot\n",
    "    df_dssp[\"p_strand\"] = df_dssp.strand / df_dssp.tot\n",
    "    df_dssp[\"p_tot\"] = df_dssp.p_helix + df_dssp.p_coil + df_dssp.p_strand\n",
    "    df_dssp.index = range(1, n + 1)\n",
    "    return df_dssp\n",
    "ss_class_df = get_ss_class_df(ss_class_dict) # CALCULATES PROPORTION OF EACH SECONDARY STRUCTURE CLASS PER POISITION\n",
    "print(len(ss_class_df))\n",
    "print(ss_class_df)\n",
    "\n",
    "# Output the secondary structure class assingments proportions as a feather \n",
    "ss_class_df.to_feather(\"/homes/2414054/Honours/231006a-AM-honours-project-aps/varalign-data-PF00026/feather/ss_class_df.feather\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reads in pickle file from VarAlign output, containing all Varaint Information\n",
    "variant_table = pd.read_pickle(r'C:\\Users\\Adam Martin\\OneDrive - University of Dundee\\Honours Project\\231006a-AM-honours-project-aps\\varalign-data-PF00026\\pickles\\PF00026.22_swissprot.sto_variants.p.gz')\n",
    "aln_in = r\"C:\\\\Users\\\\Adam Martin\\\\OneDrive - University of Dundee\\\\Honours Project\\\\231006a-AM-honours-project-aps\\\\varalign-data-PF00026\\\\alignemnt\\\\PF00026.22_swissprot.sto\" # MSA\n",
    "aln_fmt = \"stockholm\" # MSA FORMAT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source_id</th>\n",
       "      <th>protein_position</th>\n",
       "      <th>alignment_column</th>\n",
       "      <th>site</th>\n",
       "      <th>allele_num</th>\n",
       "      <th>feature</th>\n",
       "      <th>row_id</th>\n",
       "      <th>row_chrom</th>\n",
       "      <th>row_pos</th>\n",
       "      <th>row_ref</th>\n",
       "      <th>...</th>\n",
       "      <th>vep_pubmed</th>\n",
       "      <th>vep_motif_name</th>\n",
       "      <th>vep_motif_pos</th>\n",
       "      <th>vep_high_inf_pos</th>\n",
       "      <th>vep_motif_score_change</th>\n",
       "      <th>vep_lof</th>\n",
       "      <th>vep_lof_filter</th>\n",
       "      <th>vep_lof_flags</th>\n",
       "      <th>vep_lof_info</th>\n",
       "      <th>external_source_accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6618</th>\n",
       "      <td>NAPSA_HUMAN/77-401</td>\n",
       "      <td>133</td>\n",
       "      <td>154</td>\n",
       "      <td>423.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENST00000253719</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>50864990.0</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>O96009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6626</th>\n",
       "      <td>NAPSA_HUMAN/77-401</td>\n",
       "      <td>141</td>\n",
       "      <td>168</td>\n",
       "      <td>417.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENST00000253719</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>50864967.0</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>O96009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6630</th>\n",
       "      <td>NAPSA_HUMAN/77-401</td>\n",
       "      <td>145</td>\n",
       "      <td>173</td>\n",
       "      <td>413.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENST00000253719</td>\n",
       "      <td>rs746693112</td>\n",
       "      <td>19</td>\n",
       "      <td>50864954.0</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>O96009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6631</th>\n",
       "      <td>NAPSA_HUMAN/77-401</td>\n",
       "      <td>145</td>\n",
       "      <td>173</td>\n",
       "      <td>414.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENST00000253719</td>\n",
       "      <td>rs186910155</td>\n",
       "      <td>19</td>\n",
       "      <td>50864955.0</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>O96009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6639</th>\n",
       "      <td>NAPSA_HUMAN/77-401</td>\n",
       "      <td>151</td>\n",
       "      <td>180</td>\n",
       "      <td>410.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENST00000253719</td>\n",
       "      <td>rs773392864</td>\n",
       "      <td>19</td>\n",
       "      <td>50864935.0</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>O96009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41519</th>\n",
       "      <td>PEPC_HUMAN/72-387</td>\n",
       "      <td>323</td>\n",
       "      <td>646</td>\n",
       "      <td>738.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENST00000373025</td>\n",
       "      <td>rs369120707</td>\n",
       "      <td>6</td>\n",
       "      <td>41705513.0</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>P20142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41548</th>\n",
       "      <td>PEPC_HUMAN/72-387</td>\n",
       "      <td>346</td>\n",
       "      <td>719</td>\n",
       "      <td>692.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENST00000373025</td>\n",
       "      <td>None</td>\n",
       "      <td>6</td>\n",
       "      <td>41704720.0</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>P20142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41551</th>\n",
       "      <td>PEPC_HUMAN/72-387</td>\n",
       "      <td>347</td>\n",
       "      <td>720</td>\n",
       "      <td>691.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENST00000373025</td>\n",
       "      <td>rs187120335</td>\n",
       "      <td>6</td>\n",
       "      <td>41704718.0</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>P20142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41557</th>\n",
       "      <td>PEPC_HUMAN/72-387</td>\n",
       "      <td>353</td>\n",
       "      <td>728</td>\n",
       "      <td>687.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENST00000373025</td>\n",
       "      <td>rs768749667</td>\n",
       "      <td>6</td>\n",
       "      <td>41704700.0</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>P20142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41558</th>\n",
       "      <td>PEPC_HUMAN/72-387</td>\n",
       "      <td>354</td>\n",
       "      <td>729</td>\n",
       "      <td>686.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENST00000373025</td>\n",
       "      <td>rs749565140</td>\n",
       "      <td>6</td>\n",
       "      <td>41704696.0</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>P20142</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>546 rows Ã— 161 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                source_id  protein_position  alignment_column   site  \\\n",
       "6618   NAPSA_HUMAN/77-401               133               154  423.0   \n",
       "6626   NAPSA_HUMAN/77-401               141               168  417.0   \n",
       "6630   NAPSA_HUMAN/77-401               145               173  413.0   \n",
       "6631   NAPSA_HUMAN/77-401               145               173  414.0   \n",
       "6639   NAPSA_HUMAN/77-401               151               180  410.0   \n",
       "...                   ...               ...               ...    ...   \n",
       "41519   PEPC_HUMAN/72-387               323               646  738.0   \n",
       "41548   PEPC_HUMAN/72-387               346               719  692.0   \n",
       "41551   PEPC_HUMAN/72-387               347               720  691.0   \n",
       "41557   PEPC_HUMAN/72-387               353               728  687.0   \n",
       "41558   PEPC_HUMAN/72-387               354               729  686.0   \n",
       "\n",
       "       allele_num          feature       row_id row_chrom     row_pos row_ref  \\\n",
       "6618          0.0  ENST00000253719         None        19  50864990.0       T   \n",
       "6626          0.0  ENST00000253719         None        19  50864967.0       A   \n",
       "6630          0.0  ENST00000253719  rs746693112        19  50864954.0       C   \n",
       "6631          0.0  ENST00000253719  rs186910155        19  50864955.0       G   \n",
       "6639          1.0  ENST00000253719  rs773392864        19  50864935.0       G   \n",
       "...           ...              ...          ...       ...         ...     ...   \n",
       "41519         0.0  ENST00000373025  rs369120707         6  41705513.0       G   \n",
       "41548         0.0  ENST00000373025         None         6  41704720.0       C   \n",
       "41551         0.0  ENST00000373025  rs187120335         6  41704718.0       C   \n",
       "41557         0.0  ENST00000373025  rs768749667         6  41704700.0       A   \n",
       "41558         0.0  ENST00000373025  rs749565140         6  41704696.0       G   \n",
       "\n",
       "       ... vep_pubmed  vep_motif_name vep_motif_pos  vep_high_inf_pos  \\\n",
       "6618   ...                                                              \n",
       "6626   ...                                                              \n",
       "6630   ...                                                              \n",
       "6631   ...                                                              \n",
       "6639   ...                                                              \n",
       "...    ...        ...             ...           ...               ...   \n",
       "41519  ...                                                              \n",
       "41548  ...                                                              \n",
       "41551  ...                                                              \n",
       "41557  ...                                                              \n",
       "41558  ...                                                              \n",
       "\n",
       "       vep_motif_score_change  vep_lof  vep_lof_filter  vep_lof_flags  \\\n",
       "6618                                                                    \n",
       "6626                                                                    \n",
       "6630                                                                    \n",
       "6631                                                                    \n",
       "6639                                                                    \n",
       "...                       ...      ...             ...            ...   \n",
       "41519                                                                   \n",
       "41548                                                                   \n",
       "41551                                                                   \n",
       "41557                                                                   \n",
       "41558                                                                   \n",
       "\n",
       "       vep_lof_info  external_source_accession  \n",
       "6618                                    O96009  \n",
       "6626                                    O96009  \n",
       "6630                                    O96009  \n",
       "6631                                    O96009  \n",
       "6639                                    O96009  \n",
       "...             ...                        ...  \n",
       "41519                                   P20142  \n",
       "41548                                   P20142  \n",
       "41551                                   P20142  \n",
       "41557                                   P20142  \n",
       "41558                                   P20142  \n",
       "\n",
       "[546 rows x 161 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Formats the the table into only human consensus sequences with human occupancy on >50%\n",
    "def format_variant_table(df, col_mask, vep_mask = [\"missense_variant\"], tab_format = \"gnomad\"):\n",
    "    df_filt = df.copy(deep = True)\n",
    "    df_filt.reset_index(inplace=True)\n",
    "    if tab_format == \"gnomad\":\n",
    "        df_filt.columns = [' '.join(col).strip() for col in df_filt.columns.tolist()]\n",
    "    df_filt.columns = [col.lower().replace(\" \", \"_\") for col in df_filt.columns.tolist()]\n",
    "    df_filt = df_filt[df_filt.source_id.str.contains(\"HUMAN\")]\n",
    "    df_filt = df_filt.dropna(subset=[\"vep_consequence\"])\n",
    "    df_filt = df_filt[df_filt.vep_consequence.isin(vep_mask) ]\n",
    "    df_filt = df_filt[df_filt.alignment_column.isin(col_mask)]\n",
    "    return df_filt\n",
    "human_missense_variants = format_variant_table(variant_table, human_cons_cols) # GET ONLY MISSENSE VARIANTS ROWS\n",
    "human_missense_variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output the human missense variants table as a feather \n",
    "human_missense_variants.to_feather(\"/homes/2414054/Honours/231006a-AM-honours-project-aps/varalign-data-PF00026/feather/PF00026.22_swissprot.sto_variants.p.gz.feather\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
